petastorm/reader_impl/shuffling_buffer.py

```diff
@@ -0,0 +1,168 @@
+        pass
+
+    @abc.abstractmethod
+    def finish(self):
+        """Call this method when no more `enqueu_many` calls will be made.
+
+        :return: number of elements currently present in the buffer
+        """
+        pass
+
+class NoopShufflingBuffer(ShufflingBufferBase):
+    """A 'no-operation' (noop) implementation of a shuffling buffer. Useful in cases where no shuffling is desired, such
+    as test scenarios or iterating over a dataset in a predeterministic order.
+    """
+    def __init__(self):
+        self.store = []
+
+    def add_many(self, items):
+        self.store.extend(items)
+
+    def retrieve(self):
+        return self.store.pop(0)
+
+    def can_retrieve(self):
+        return len(self.store) > 0
+
+    def can_add(self):
+        return True
+
+    @property
+    def size(self):
+        return len(self.store)
+
+    def finish(self):
+        pass
+
+class RandomShufflingBuffer(ShufflingBufferBase):
+    """
+    A random shuffling buffer implementation. Items can be added to the buffer and retrieved in a random order.
+    """
+    def __init__(self, shuffling_buffer_capacity, min_after_retrieve, extra_capacity=1000):
+        """Initializes a new ShufflingBuffer instance.
+
+        Items may be retrieved from the buffer once `min_after_dequeue` items were added to the queue
+        (indicated by `can_retrieve`).
+
+        Items may be added to the buffer as long as the amount of items in the buffer does not exceed
+        `shuffling_queue_capacity`.
+
+        The amount of items in the buffer may grow above `shuffling_buffer_capacity` (due to a call to `add_many`
+        with a list of items), but must remain under `extra_queue_capacity`.
+
+        :param shuffling_buffer_capacity: Items may be added to the buffer as long as the amount of items in the
+        buffer does not exceed the value of `shuffling_queue_capacity`.
+        :param min_after_retrieve: The amount of items in the buffer may grow above `shuffling_buffer_capacity`
+        (due to a call to `add_many` with a list of items), but must remain under `extra_queue_capacity`
+        """
+        self._extra_capacity = extra_capacity
+        # Preallocate the shuffling buffer.
+        self._items = [None] * (shuffling_buffer_capacity + self._extra_capacity)
+        self._shuffling_queue_capacity = shuffling_buffer_capacity
+        self._min_after_dequeue = min_after_retrieve
+        self._size = 0
+        self._done_adding = False
+
+    def add_many(self, items):
+        if self._done_adding:
+            raise RuntimeError('Can not call add_many after done_adding() was called.')
+
+        if not self.can_add():
+            raise RuntimeError('Can not enqueue. Check the return value of "can_enqueue()" to check if more '
+                               'items can be added.')
+
+        # We leave self._extra_capacity slack to make sure we don't reallocate self._items array
+        expected_size = self.size + len(items)
```

Comment:
the mix of using `self.size` or `self._size` is a bit confusing. Perhaps stick to `self._size`

Reply:
Done